src/k8s/pkg/k8sd/api/certs_refresh.go

```diff
@@ -0,0 +1,346 @@
+	}
+
+	var seed int32
+
+	err = binary.Read(rand.Reader, binary.BigEndian, &seed)
+	if err != nil {
+		return response.InternalError(err)
+	}
+	seed = seed & 0x7FFFFFFF
+
+	log.Info("Generating CSRs for worker node")
+	log.Info("Generating Kubelet Serving Certificate Signing Request")
+
+	csrKubeletServing, pKeyKubeletServing, err := pkiutil.GenerateCSR(
+		pkix.Name{
+			CommonName:   fmt.Sprintf("system:node:%s", snap.Hostname()),
+			Organization: []string{"system:nodes"},
+		},
+		2048,
+		nil,
+		[]string{snap.Hostname()},
+		[]net.IP{net.ParseIP(s.Address().Hostname())},
+	)
+
+	_, err = client.CreateCertificateSigningRequest(
+		r.Context(),
+		fmt.Sprintf("k8sd-%d-worker-kubelet-serving", seed),
+		[]byte(csrKubeletServing),
+		[]v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageServerAuth},
+		[]string{"system:nodes"},
+		"k8sd.io/kubelet-serving",
+	)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	log.Info("Generating Kubelet Client Certificate Signing Request")
+	csrKubeletClient, pKeyKubeletClient, err := pkiutil.GenerateCSR(
+		pkix.Name{
+			CommonName:   fmt.Sprintf("system:node:%s", snap.Hostname()),
+			Organization: []string{"system:nodes"},
+		},
+		2048,
+		nil,
+		nil,
+		nil,
+	)
+
+	_, err = client.CreateCertificateSigningRequest(
+		r.Context(),
+		fmt.Sprintf("k8sd-%d-worker-kubelet-client", seed),
+		[]byte(csrKubeletClient),
+		[]v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageClientAuth},
+		nil,
+		"k8sd.io/kubelet-client",
+	)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	log.Info("Generating Kube Proxy Client Certificate Signing Request")
+	csrKubeProxy, pKeyKubeProxy, err := pkiutil.GenerateCSR(
+		pkix.Name{
+			CommonName: "system:kube-proxy",
+		},
+		2048,
+		nil,
+		nil,
+		nil,
+	)
+
+	_, err = client.CreateCertificateSigningRequest(
+		r.Context(),
+		fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", seed),
+		[]byte(csrKubeProxy),
+		[]v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageClientAuth},
+		nil,
+		"k8sd.io/kube-proxy-client",
+	)
+
+	result := apiv1.RefreshCertificatesPlanResponse{
+		Seed: int(seed),
+		CertificatesSigningRequests: []string{
+			fmt.Sprintf("k8sd-%d-worker-kubelet-serving", seed),
+			fmt.Sprintf("k8sd-%d-worker-kubelet-client", seed),
+			fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", seed),
+		},
+	}
+
+	log.Info("Writing new private keys")
+	operations := []utils.FileOperations{
+		{
+			SourcePath:  filepath.Join(snap.KubernetesPKIDir(), "kubelet.key.tmp"),
+			Content:     []byte(pKeyKubeletServing),
+			Permissions: 0600,
+		},
+		{
+			SourcePath:  filepath.Join(snap.KubernetesPKIDir(), "kubelet-client.key.tmp"),
+			Content:     []byte(pKeyKubeletClient),
+			Permissions: 0600,
+		},
+		{
+			SourcePath:  filepath.Join(snap.KubernetesPKIDir(), "kube-proxy.key.tmp"),
+			Content:     []byte(pKeyKubeProxy),
+			Permissions: 0600,
+		},
+	}
```

Comment:
Note: `/plan` should only return the seed. The CSR creation should happen in `/run`, and private keys do not need to be shared by writing them on the filesystem.

Reply:
I'm confused. In the [proposal](https://github.com/canonical/k8s-snap/blob/main/docs/proposals/002-refresh-certificates.md#post-refresh-certsplan) I see that for the `plan` endpoint it should return the CSRs list for the worker nodes:

> It returns a seed that must be passed to POST /refresh-certs/run. For worker nodes, it also returns a list of names of CertificateSigningRequest objects that will need to be approved and signed. This can be used by the CLI to print a helpful message and request that the certificates be signed.

So I believe the CSR create should happen in `/plan` instead of `/run`. Am I missing something?

Reply:
OH, I see. This is the heart of the matter:

> This method call is purely informative and does not have any side-effect on the cluster.

So I think the meaning here:

> POST /refresh-certs/plan performs any necessary preparations for certificate refreshes on the current node.

is that these preparations are only to generate the seed and the names for the CSRs that will be created in the future during the `run` endpoint execution.